### ISSUE TITLE
generateChangeLog not working

### ISSUE THREAD

**Issue by Magn1fico** [liquibase/liquibase-neo4j#285]
Hello everyone, 

I'm making first steps in liquibase plugin for neo4j and I'm struggling with generating the initial changeLog file for the existing neo4j database. 

I run the following command:
`liquibase --changeLogFile=mydatabase_changelog.xml generateChangeLog`

The error I get is:
`Unexpected error running Liquibase: Don't know how to query for sequences on neo4j @ jdbc:neo4j:neo4j+s://_databaseaddress_`

At the same time, I am able to connect to the database and run a simple changelog which creates a node as per quick start guide, so my guess it is not a connectivity issue. 

Can anyone adivse, please, if generateChangeLog functionality working with neo4j?

**Comment by fbiville** [comment]
Hello, thanks for the report, I'll need to estimate how wide the gap is to support that functionality.
I'll let you know ASAP.

**Comment by fbiville** [comment]
Hello, @mgazanayi and I had a look and it appears `generateChangeLog` will not be easy to support due to the nature of Neo4j.

**TL;DR:** the best we can do is generate the index/constraint changes to match those detected in the database. That needs to happen after we implement the `createIndex`/`createConstraint` changes in the plugin.

**Long version ⬇️**

Relational databases have a strict separation of structure and data.
You must first define structure (table, columns) before inserting data into it. These structures can be reliably retrieved and that's what `generateChangeLog` does: get the existing structures and generate the corresponding Liquibase change for it.

With Neo4j, there's no strict separation between structure and data. They are tied together.
You create the data directly (with the exception of indices and constraints).

I don't think it would be reasonable implementation to dump the whole data into a Liquibase change set as a result of `generateChangeLog`. The data set could be huge, leading to memory issues and probably for no added value (unless each node/relationship has its own schema - which is not true, most of the time).

If we exclude data, the only artefacts we have left are indices & constraints.
We could inspect the target Neo4j database and generate the Liquibase changes that would in turn create indices & constraints when run.

Currently, this is done with raw Cypher queries. @mgazanayi and I just started looking into implementing higher-level constructs that support indices/constraints creation & drop in a Neo4j version-agnostic manner. We want to focus on that first.

**Comment by Magn1fico** [comment]
Hello @fbiville! 

Thank you very much for looking into my question, and for such a detailed explanation. 

At least I am now sure that this is not implemented, and not me doing something wrong. 

I guess I'll just repeat the data modeling steps on a new Neo4j instance from scratch. 

Cheers!

**Comment by fbiville** [comment]
If you can share of your manual work here, that would be super insightful for us!
We would love to learn more about the use cases of `generateChangelog` for Neo4j.

**Comment by rberezen** [comment]
This was well explained by @fbiville — `generateChangeLog` for Neo4j is a design limitation due to the nature of graph databases. The user confirmed they understood. Closing as this is a known limitation, not a bug. Please reopen if there's new context or if index/constraint generation support is needed.
